## 練習問題

by kenjisato  
id: ch01-kjst03

### 問題



(ch01.pdf, 1.3.4 から転載)

次の5つのコードを実行したときに表示されるエラーメッセージに目を通し，「エラー名：エラーの理由」を書き取りなさい。エラーの原因について，自分の言葉で説明しなさい。

1. `3x = 10`
2. `True = 0`
3. `a-b = 0`
4. `print(abcde)`
5. `{3 * (2 + 4)} * 3`

In [1]:
3x = 10

SyntaxError: invalid syntax (<ipython-input-1-2fa14a371807>, line 1)

`SyntaxError: invalid syntax`

文法が間違えている。`3x` という名前の変数を作りたかったのだけれど，数字からはじまる名前は禁止である。代入される前にダメな名前が見つかったから文法エラーになる。

In [3]:
True = 0

SyntaxError: can't assign to keyword (<ipython-input-3-5a198676a762>, line 1)

`True` とか `False` はキーワード（予約語）と言われるもので，代入できない。

In [4]:
a-b = 0

SyntaxError: can't assign to operator (<ipython-input-4-7d6daa96765b>, line 1)

`SyntaxError: can't assign to operator`

代入する文の左辺には演算子（operator, ここでは `-`）を書くことはできない。

In [5]:
print(abcde)

NameError: name 'abcde' is not defined

`NameError: name 'abcde' is not defined`

`abcde` が定義されていない。

In [6]:
{3 * (2 + 4)} * 3

TypeError: unsupported operand type(s) for *: 'set' and 'int'

`TypeError: unsupported operand type(s) for *: 'set' and 'int'`

これはちょっと難しい。`set` (集合) オブジェクトと `int` (整数) オブジェクトの乗算 `*` はサポートされていない。`*` がサポートしていない Type (型) の演算をやろうとしたのでエラーになっている。

### 集合

ちなみに，`set` というのは次のように使う。集合 $s = \{1, 2, 3\}$ と $t = \{3, 4, 5\}$ の和集合 $s \cup t = \{1, 2, 3, 4, 5\}$ を表示してみよう。

In [7]:
s = {1, 2, 3}
s

{1, 2, 3}

In [8]:
t = {3, 4, 5}
t

{3, 4, 5}

In [9]:
s.union(t)

{1, 2, 3, 4, 5}

集合に要素が入っているかの判定は `in` を使ってできる。

In [10]:
1 in s

True

In [11]:
1 in t

False

集合はリストのように見えるが使いみちが違う。一番大きな違いは

- リストは順序付き
- 集合は順序なし

ということ，順序付けをしないことで，集合に対しては高速な要素の検索が可能になっている。

In [12]:
l = list(range(100000))
%timeit 99999 in l

1.08 ms ± 67.3 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)


In [13]:
s = set(l)
%timeit 99999 in s

45.8 ns ± 2.46 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)


`%timeit` はコードの実行時間を測る IPython のマジック関数である。リストに対する検索は平均 1.08ミリ秒，集合に対しては 45.8ナノ秒なので，集合を使うほうが 23.8倍くらい早い（このファクターは検索対象の大きさにも依存する）。